lawnchair/src/app/lawnchair/search/GenerateSearchTarget.kt

```diff
@@ -0,0 +1,198 @@
+        val fileIntent = Intent(Intent.ACTION_VIEW)
+            .addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
+            .addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
+            .setDataAndType(info.path.path2Uri(), info.mimeType.mimeCompat)
+
+        val action = SearchActionCompat.Builder(info.path, info.name)
+            .setIcon(getPreviewIcon(info))
+            .setIntent(fileIntent)
+            .build()
+
+        return createSearchTarget(
+            info.path,
+            action,
+            LayoutType.THUMBNAIL,
+            SearchTargetCompat.RESULT_TYPE_FILE_TILE,
+            FILES,
+            Bundle(),
+        )
+    }
+
+    private fun displayContactPhoto(context: Context, name: String, contactUri: Uri): Icon {
+        var inputStream: InputStream? = null
+        try {
+            inputStream = context.contentResolver.openInputStream(contactUri)
+
+            inputStream?.use { input ->
+                val bitmap = BitmapFactory.decodeStream(input)
+                return Icon.createWithAdaptiveBitmap(bitmap)
+            }
+        } catch (e: IOException) {
+            // ignore
+        } finally {
+            inputStream?.close()
+        }
+
+        // If contact photo not available, create an icon with the first letter of the name
+        val initial = if (name.isNotEmpty()) name[0].uppercaseChar().toString() else "U"
+        val textBitmap = createTextBitmap(context, initial)
+        return Icon.createWithBitmap(textBitmap)
+    }
+
+    private fun getStartPageUrl(query: String): String {
+        return "https://www.startpage.com/do/search?segment=startpage.lawnchair&query=$query&cat=web"
+    }
+
+    private fun resolveMarketSearchActivity(): ComponentKey? {
+        val intent = PackageManagerHelper.getMarketSearchIntent(context, "")
+        val resolveInfo = context.packageManager.resolveActivity(intent, 0) ?: return null
+        val packageName = resolveInfo.activityInfo.packageName
+        val launchIntent =
+            context.packageManager.getLaunchIntentForPackage(packageName) ?: return null
+        return ComponentKey(launchIntent.component, Process.myUserHandle())
+    }
+
+    private fun getPreviewIcon(info: FileInfo): Icon = when {
+        info.isMediaType -> BitmapFactory.decodeFile(info.path)?.let { Icon.createWithBitmap(it) }
+            ?: (
+                if (Utilities.ATLEAST_R) {
+                    MediaMetadataRetriever().run {
+                        setDataSource(info.path)
+                        val videoBitmap = frameAtTime
+                        release()
+                        videoBitmap?.let { Icon.createWithBitmap(it) }
+                    }
+                } else {
+                    null
+                }
+                ) ?: Icon.createWithResource(context, info.iconRes)
+        info.isUnknownType -> Icon.createWithBitmap(createTextBitmap(context, "U"))
+        else -> Icon.createWithResource(context, info.iconRes)
```

Comment:
We can defer this bitmap fetching, just pass file path as the icon param, then load it with Coil.